zipper/protocols/querydispatch/querydispatch_helper.go

```diff
@@ -0,0 +1,255 @@
+package querydispatch
+
+import (
+	"fmt"
+
+	"github.com/go-graphite/carbonapi/zipper/broadcast"
+	"github.com/go-graphite/carbonapi/zipper/metadata"
+	"github.com/go-graphite/carbonapi/zipper/types"
+	protov3 "github.com/go-graphite/protocol/carbonapi_v3_pb"
+	"go.uber.org/zap"
+)
+
+func (q *QueryDispatchGroup) CreatePerFetchQueryStoreBackends(request *protov3.MultiFetchRequest) (map[types.BackendServer][]*protov3.FetchRequest, error) {
+	// Match each regex rule to each query and put backend in allow or deny groups.
+	AllowGroupPerMetric := make(map[*protov3.FetchRequest][]string)
+	DenyGroupPerMetric := make(map[*protov3.FetchRequest][]string)
+	for _, fetchReq := range request.GetMetrics() {
+		metricReq := fetchReq.GetName()
+		for _, rule := range q.backendsV2.GroupRegexRules {
+			if rule.CompiledRegex.MatchString(metricReq) {
+				if rule.Effect == "allow" {
+					AllowGroupPerMetric[&fetchReq] = types.AppendUnique(AllowGroupPerMetric[&fetchReq], rule.GroupBackendNames)
+				} else if rule.Effect == "deny" {
+					DenyGroupPerMetric[&fetchReq] = types.AppendUnique(DenyGroupPerMetric[&fetchReq], rule.GroupBackendNames)
+				} else {
+					q.logger.Warn("Unknow Effect")
+				}
+			}
+		}
+	}
+	// If a backend is denied somewhere it has to be removed from the allow list
+	for query, allowedGroupList := range AllowGroupPerMetric {
+		denyGroupList, ok := DenyGroupPerMetric[query]
+		if !ok {
+			continue
+		}
+		for index, allowedGroupName := range allowedGroupList {
+			if contains(denyGroupList, allowedGroupName) {
+
+				remove(AllowGroupPerMetric[query], index)
+				continue
+			}
+		}
+	}
+	// Reverse Order
+	queriesPerGroup := make(map[string][]*protov3.FetchRequest)
+	for query, allowedGroupList := range AllowGroupPerMetric {
+		for _, nameGroup := range allowedGroupList {
+			queriesPerGroup[nameGroup] = append(queriesPerGroup[nameGroup], query)
+		}
+	}
+
```

Comment:
What happens if there is no match? Shouldn't we broadcast or use a user supplied strategy?